kawin/Diffusion.py

```diff
@@ -0,0 +1,825 @@
+        '''
+        Sets composition as a step-wise function
+
+        Parameters
+        ----------
+        Lvalue : float
+            Value on left side of mesh
+        Rvalue : float
+            Value on right side of mesh
+        z : float
+            Position on mesh where composition switches from Lvalue to Rvalue
+        element : str
+            Element to apply composition profile to
+        '''
+        eIndex = self._getElementIndex(element)
+        Lindices = self.z <= z
+        self.x[eIndex,Lindices] = Lvalue
+        self.x[eIndex,~Lindices] = Rvalue
+
+    def setCompositionSingle(self, value, z, element = None):
+        '''
+        Sets single node to specified composition
+
+        Parameters
+        ----------
+        value : float
+            Composition
+        z : float
+            Position to set value to (will use closest node to z)
+        element : str
+            Element to apply composition profile to
+        '''
+        eIndex = self._getElementIndex(element)
+        zIndex = np.argmin(np.abs(self.z-z))
+        self.x[eIndex,zIndex] = value
+
+    def setCompositionInBounds(self, value, Lbound, Rbound, element = None):
+        '''
+        Sets single node to specified composition
+
+        Parameters
+        ----------
+        value : float
+            Composition
+        Lbound : float
+            Position of left bound
+        Rbound : float
+            Position of right bound
+        element : str
+            Element to apply composition profile to
+        '''
+        eIndex = self._getElementIndex(element)
+        indices = (self.z >= Lbound) & (self.z <= Rbound)
+        self.x[eIndex,indices] = value
+
+    def setCompositionFunction(self, func, element = None):
+        '''
+        Sets composition as a function of z
+
+        Parameters
+        ----------
+        func : function
+            Function taking in z and returning composition
+        element : str
+            Element to apply composition profile to
+        '''
+        eIndex = self._getElementIndex(element)
+        self.x[eIndex,:] = func(self.z)
+
+    def setCompositionProfile(self, z, x, element = None):
+        '''
+        Sets composition profile by linear interpolation
+
+        Parameters
+        ----------
+        z : array
+            z-coords of composition profile
+        x : array
+            Composition profile
+        element : str
+            Element to apply composition profile to
+        '''
+        eIndex = self._getElementIndex(element)
+        z = np.array(z)
+        x = np.array(x)
+        sortIndices = np.argsort(z)
+        z = z[sortIndices]
+        x = x[sortIndices]
+        self.x[eIndex,:] = np.interp(self.z, z, x)
+
+    def setup(self):
+        self.therm.clearCache()
+        xsum = np.sum(self.x, axis=0)
+        if any(xsum > 1):
+            raise Exception('Warning: some compositions sum up to above 1')
```

Comment:
Warnings are different from exceptions in Python. I think it's clearer to remove 'Warning' from the exception message.

Reply:
Line 309 will be changed to 
`raise Exception('Some compositions sum up to above 1')`

I will also work on adding a line that says where in the mesh the compositions add up to over 1.